### PR TITLE
Product

### PR DESCRIPTION
**Pull request by Mr-Rovilay**
Check the product page

### CODE REVIEW

File: server/controllers/productController.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    res.status(200).json(product);
+  } catch (error) {
+    console.error("Error fetching menu item:", error);
+    res.status(500).json({ error: "Internal server error" });
+  }
+};
+
+export const getAllProduct = async (req, res) => {
+  const New = req.query.new;
```
Comment (rovilay) [inline server/controllers/productController.js]: Good idea, but sort (ascending or descending) is better.

Comment (rovilay) [reply server/controllers/productController.js]: Also, do not use capital letters

File: server/controllers/productController.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  const Category = req.query.category;
+  try {
+    let product;
+    if (New) {
+      product = await Product.find().sort({ createdAt: -1 }).limit(2);
```
Comment (rovilay) [inline server/controllers/productController.js]: incorporate pagination instead of hardcoding limit value

File: server/routes/productRoutes.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+} from "../controllers/productController.js";
+import { verifyTokenAndAdmin } from "../middleware/verifyJWT.js";
+import { validateProduct } from "../middleware/validate.js";
+
+const router = express.Router();
+
+router.post("/", verifyTokenAndAdmin, validateProduct, createPost);
+router.get("/:id", getProduct);
+router.get("/", getAllProduct);
+router.delete("/:id", verifyTokenAndAdmin, deletedProduct);
```
Comment (rovilay) [inline server/routes/productRoutes.js]: update name to `deleteProduct`

File: server/middleware/validate.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+export const validateProduct = [
+  check("title").not().isEmpty().withMessage("Title is required"),
+  check("price")
+    .isFloat({ gt: 0 })
+    .withMessage("Price must be a positive number"),
+  check("description").not().isEmpty().withMessage("Description is required"),
+  check("categories")
+    .optional()
+    .isArray()
+    .withMessage("Categories must be an array"),
```
Comment (rovilay) [inline server/middleware/validate.js]: validate categories

File: server/middleware/verifyJWT.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  const authHeader = req.headers.token;
+
+  if (!authHeader) {
+    return res.status(401).send({ error: "Unauthorized" });
+  }
+
+  const token = authHeader.split(" ")[1];
+  jwt.verify(token, process.env.SECRET_ACCESS_KEY, (err, user) => {
+    if (err) {
+      return res.status(403).send({ error: "Invalid token" });
```
Comment (rovilay) [inline server/middleware/verifyJWT.js]: should be 401

File: server/middleware/verifyJWT.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+export const verifyTokenAndAuthorization = (req, res, next) => {
+  verifyJWT(req, res, () => {
+    if (req.decoded.id === req.params.id || req.decoded.isAdmin) {
+      next();
+    } else {
+      return res.status(403).send({ error: "You are not authorized" });
+    }
+  });
+};
```
Comment (rovilay) [inline server/middleware/verifyJWT.js]: Logic is wrong please review. Check req.decoded is not undefined instead or you can remove and use the verifyJWT instead.

File: server/models/productModel.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    title: {
+      type: String,
+      required: true,
+      unique: true,
+    },
+    description: {
+      type: String,
+      required: true,
+    },
+    image: { type: String, required: true },
```
Comment (rovilay) [inline server/models/productModel.js]: verify image is url

File: server/controllers/userControllers.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+//get all users
+export const getAllUsers = async (req, res) => {
+  const query = req.query.new;
+  try {
+    const users = query
+      ? await User.find().sort({ _id: -1 }).limit(5)
+      : await User.find().select("-password");
+    res.status(200).json(users);
```
Comment (rovilay) [inline server/controllers/userControllers.js]: add sort query and pagination.

File: server/index.js [hunk 9]
```diff
@@ -13,6 +15,8 @@ app.use(express.urlencoded({ extended: true }));
 app.use(express.json({ limit: "50mb" }));
 
 app.use("/auth", authRoutes);
+app.use("/user", userRoutes);
+app.use("/product", productRoutes);
```
Comment (rovilay) [inline server/index.js]: pluralize your routes e.g `/products`, `/users`

**Review by rovilay — CHANGES_REQUESTED** [review]
Good work but implement the feedback.

File: server/controllers/productController.js (hunk 1 above)
Comment (Mr-Rovilay) [reply server/controllers/productController.js]: done.. needs review

File: server/controllers/productController.js (hunk 2 above)
Comment (Mr-Rovilay) [reply server/controllers/productController.js]: done.. needs review

File: server/routes/productRoutes.js (hunk 3 above)
Comment (Mr-Rovilay) [reply server/routes/productRoutes.js]: done.. needs review

File: server/middleware/validate.js (hunk 4 above)
Comment (Mr-Rovilay) [reply server/middleware/validate.js]: done.. needs review

File: server/middleware/verifyJWT.js (hunk 5 above)
Comment (Mr-Rovilay) [reply server/middleware/verifyJWT.js]: done.. needs review

File: server/middleware/verifyJWT.js (hunk 6 above)
Comment (Mr-Rovilay) [reply server/middleware/verifyJWT.js]: done.. needs review

File: server/models/productModel.js (hunk 7 above)
Comment (Mr-Rovilay) [reply server/models/productModel.js]: done.. needs review

File: server/controllers/userControllers.js (hunk 8 above)
Comment (Mr-Rovilay) [reply server/controllers/userControllers.js]: done.. needs review

File: server/index.js (hunk 9 above)
Comment (Mr-Rovilay) [reply server/index.js]: done.. needs review

**Pull request merged**